classes/course_file.php

```diff
@@ -201,122 +204,107 @@ protected function get_file_download_url($file) {
+    protected function get_standard_file_download_url(bool $insertitemid = true) : \moodle_url {
+        $url = '/pluginfile.php/' . $this->file->contextid . '/' . $this->file->component . '/' . $this->file->filearea;
         if ($insertitemid) {
-            $url .= '/' . $file->itemid;
+            $url .= '/' . $this->file->itemid;
         }
-        $url .= $file->filepath . $file->filename;
-
+        $url .= $this->file->filepath . $this->file->filename;
         return new \moodle_url($url);
     }
 
     /**
      * Try to get the url for the component (module or course).
      *
-     * @param object $file
      * @return null|\moodle_url
-     * @throws moodle_exception
+     * @throws \moodle_exception
      */
-    protected function get_component_url($file) {
-        if ($file->contextlevel == CONTEXT_MODULE) {
-            $this->coursemodinfo = get_fast_modinfo($this->courseid);
-            if (!empty($this->coursemodinfo->cms[$file->instanceid])) {
-                return $this->coursemodinfo->cms[$file->instanceid]->url;
+    protected function get_component_url() : ?\moodle_url {
+        if ($this->file->contextlevel == CONTEXT_MODULE) {
+            $coursemodinfo = get_fast_modinfo($this->courseid);
+            if (!empty($coursemodinfo->cms[$this->file->instanceid])) {
+                return $coursemodinfo->cms[$this->file->instanceid]->url;
             }
         }
-
         return null;
     }
 
     /**
      * Checks if embedded files have been used
      *
-     * @param object $file
      * @return bool|null
      * @throws \dml_exception
      */
-    protected function is_file_used($file) {
+    protected function is_file_used() : ?bool {
         global $DB;
-        $isused = null;
-        $component = strpos($file->component, 'mod_') === 0 ? 'mod' : $file->component;
-
+        $component = strpos($this->file->component, 'mod_') === 0 ? 'mod' : $this->file->component;
         switch ($component) {
-            case 'mod' : // Course module.
-                $modname = str_replace('mod_', '', $file->component);
-                if ($file->filearea === 'intro') {
-                    $sql = 'SELECT m.* FROM {context} ctx
-                            JOIN {course_modules} cm ON cm.id = ctx.instanceid
-                            JOIN {' . $modname . '} m ON m.id = cm.instance
-                            WHERE ctx.id = ?';
-                    $mod = $DB->get_record_sql($sql, [$file->contextid]);
-                    $isused = $this->is_embedded_file_used($mod, 'intro', $file->filename);
+            case 'mod': // Course module.
+                $modname = str_replace('mod_', '', $this->file->component);
+                if ($this->file->filearea === 'intro') {
+                    $sql = "SELECT m.*
+                              FROM {context} ctx
+                              JOIN {course_modules} cm ON cm.id = ctx.instanceid
+                              JOIN {$modname} m ON m.id = cm.instance
```

Comment:
Remember that curly braces have a special meaning in double quoted strings and Moodle needs the curly braces around table names.
To be sure that sql injections are definitely impossible, please check whether `$modname` is the name of an installed plugin. `core_component::get_plugin_list('mod')`

Reply:
You're right of course. Sorry about that oversight. Should be fixed now. I also added the additional check you requested. Hope it works as expected now.